File: c/src/realestate.c
/*
 * realestate.c — CircleAI.RealEstate (C11 port of RealEstatePrimitives.cs).
 *
 * InMemoryRealEstateBoard: properties (PropertyId keyed), listings (ListingId
 * keyed), valuations + viewings (flat append lists). ActiveInSuburb joins a
 * listing to its property's suburb; SuburbAverage means the asking prices.
 * Pure C11 + libc. No pthreads.
 */

#include "circle_ai/realestate.h"
#include "board_common.h"

/* ── record deep-copy / free ────────────────────────────────────────────── */

void ca_re_property_free(ca_re_property_t *p) {
    if (!p) return;
    free(p->property_id);
    free(p->suburb);
    p->property_id = p->suburb = NULL;
}

static bool property_copy(ca_re_property_t *dst, const ca_re_property_t *src) {
    memset(dst, 0, sizeof(*dst));
    dst->property_id   = cab_strdup_empty(src->property_id);
    dst->suburb        = cab_strdup_empty(src->suburb);
    dst->kind          = src->kind;
    dst->beds          = src->beds;
    dst->baths         = src->baths;
    dst->floor_area_m2 = src->floor_area_m2;
    if (!dst->property_id || !dst->suburb) { ca_re_property_free(dst); return false; }
    return true;
}

void ca_re_listing_free(ca_re_listing_t *l) {
    if (!l) return;
    free(l->listing_id);
    free(l->property_id);
    free(l->currency);
    l->listing_id = l->property_id = l->currency = NULL;
}
void ca_re_listing_free_array(ca_re_listing_t *arr, size_t count) {
    if (!arr) return;
    for (size_t i = 0; i < count; ++i) ca_re_listing_free(&arr[i]);
    free(arr);
}

static bool listing_copy(ca_re_listing_t *dst, const ca_re_listing_t *src) {
    memset(dst, 0, sizeof(*dst));
    dst->listing_id    = cab_strdup_empty(src->listing_id);
    dst->property_id   = cab_strdup_empty(src->property_id);
    dst->currency      = cab_strdup_empty(src->currency);
    dst->asking_price  = src->asking_price;
    dst->listed_utc_ms = src->listed_utc_ms;
    dst->is_active     = src->is_active;
    if (!dst->listing_id || !dst->property_id || !dst->currency) {
        ca_re_listing_free(dst);
        return false;
    }
    return true;
}

void ca_re_valuation_free(ca_re_valuation_t *v) {
    if (!v) return;
    free(v->property_id);
    free(v->source);
    v->property_id = v->source = NULL;
}

static bool valuation_copy(ca_re_valuation_t *dst,
                           const ca_re_valuation_t *src) {
    memset(dst, 0, sizeof(*dst));
    dst->property_id     = cab_strdup_empty(src->property_id);
    dst->source          = cab_strdup_empty(src->source);
    dst->estimated_value = src->estimated_value;
    dst->at_utc_ms       = src->at_utc_ms;
    if (!dst->property_id || !dst->source) { ca_re_valuation_free(dst); return false; }
    return true;
}

void ca_re_viewing_free(ca_re_viewing_t *v) {
    if (!v) return;
    free(v->viewing_id);
    free(v->listing_id);
    free(v->attendee_name);
    v->viewing_id = v->listing_id = v->attendee_name = NULL;
}

static bool viewing_copy(ca_re_viewing_t *dst, const ca_re_viewing_t *src) {
    memset(dst, 0, sizeof(*dst));
    dst->viewing_id    = cab_strdup_empty(src->viewing_id);
    dst->listing_id    = cab_strdup_empty(src->listing_id);
    dst->attendee_name = cab_strdup_empty(src->attendee_name);
    dst->at_utc_ms     = src->at_utc_ms;
    if (!dst->viewing_id || !dst->listing_id || !dst->attendee_name) {
        ca_re_viewing_free(dst);
        return false;
    }
    return true;
}

/* ── board ──────────────────────────────────────────────────────────────── */

struct ca_re_board {
    ca_re_property_t  *props;
    size_t             p_count, p_cap;
    ca_re_listing_t   *listings;
    size_t             l_count, l_cap;
    ca_re_valuation_t *vals;
    size_t             val_count, val_cap;
    ca_re_viewing_t   *viewings;
    size_t             view_count, view_cap;
};

ca_re_board_t *ca_re_board_create(void) {
    return (ca_re_board_t *)calloc(1, sizeof(ca_re_board_t));
}
void ca_re_board_destroy(ca_re_board_t *b) {
    if (!b) return;
    for (size_t i = 0; i < b->p_count; ++i)    ca_re_property_free(&b->props[i]);
    for (size_t i = 0; i < b->l_count; ++i)    ca_re_listing_free(&b->listings[i]);
    for (size_t i = 0; i < b->val_count; ++i)  ca_re_valuation_free(&b->vals[i]);
    for (size_t i = 0; i < b->view_count; ++i) ca_re_viewing_free(&b->viewings[i]);
    free(b->props);
    free(b->listings);
    free(b->vals);
    free(b->viewings);
    free(b);
}

int ca_re_board_register_property(ca_re_board_t *b, const ca_re_property_t *p) {
    if (!b || !p) return -1;
    for (size_t i = 0; i < b->p_count; ++i) {
        if (cab_ord_eq(b->props[i].property_id, p->property_id)) {
            ca_re_property_t copy;
            if (!property_copy(&copy, p)) return -1;
            ca_re_property_free(&b->props[i]);
            b->props[i] = copy;
            return 0;
        }
    }
    ca_re_property_t copy;
    if (!property_copy(&copy, p)) return -1;
    if (b->p_count == b->p_cap) {
        size_t nc = b->p_cap ? b->p_cap * 2 : 4;
        void *n = realloc(b->props, nc * sizeof(*b->props));
        if (!n) { ca_re_property_free(&copy); return -1; }
        b->props = (ca_re_property_t *)n;
        b->p_cap = nc;
    }
    b->props[b->p_count++] = copy;
    return 0;
}

int ca_re_board_list(ca_re_board_t *b, const ca_re_listing_t *l) {
    if (!b || !l) return -1;
    for (size_t i = 0; i < b->l_count; ++i) {
        if (cab_ord_eq(b->listings[i].listing_id, l->listing_id)) {
            ca_re_listing_t copy;
            if (!listing_copy(&copy, l)) return -1;
            ca_re_listing_free(&b->listings[i]);
            b->listings[i] = copy;
            return 0;
        }
    }
    ca_re_listing_t copy;
    if (!listing_copy(&copy, l)) return -1;
    if (b->l_count == b->l_cap) {
        size_t nc = b->l_cap ? b->l_cap * 2 : 4;
        void *n = realloc(b->listings, nc * sizeof(*b->listings));
        if (!n) { ca_re_listing_free(&copy); return -1; }
        b->listings = (ca_re_listing_t *)n;
        b->l_cap = nc;
    }
    b->listings[b->l_count++] = copy;
    return 0;
}

int ca_re_board_close(ca_re_board_t *b, const char *listing_id) {
    if (!b || !listing_id) return -1;
    for (size_t i = 0; i < b->l_count; ++i) {
        if (cab_ord_eq(b->listings[i].listing_id, listing_id)) {
            b->listings[i].is_active = false;
            return 0;
        }
    }
    return 1; /* InvalidOperationException: unknown listing */
}

int ca_re_board_value(ca_re_board_t *b, const ca_re_valuation_t *v) {
    if (!b || !v) return -1;
    ca_re_valuation_t copy;
    if (!valuation_copy(&copy, v)) return -1;
    if (b->val_count == b->val_cap) {
        size_t nc = b->val_cap ? b->val_cap * 2 : 4;
        void *n = realloc(b->vals, nc * sizeof(*b->vals));
        if (!n) { ca_re_valuation_free(&copy); return -1; }
        b->vals = (ca_re_valuation_t *)n;
        b->val_cap = nc;
    }
    b->vals[b->val_count++] = copy;
    return 0;
}

int ca_re_board_schedule_viewing(ca_re_board_t *b, const ca_re_viewing_t *v) {
    if (!b || !v) return -1;
    ca_re_viewing_t copy;
    if (!viewing_copy(&copy, v)) return -1;
    if (b->view_count == b->view_cap) {
        size_t nc = b->view_cap ? b->view_cap * 2 : 4;
        void *n = realloc(b->viewings, nc * sizeof(*b->viewings));
        if (!n) { ca_re_viewing_free(&copy); return -1; }
        b->viewings = (ca_re_viewing_t *)n;
        b->view_cap = nc;
    }
    b->viewings[b->view_count++] = copy;
    return 0;
}

/* Does a listing's property match `suburb` (OrdinalIgnoreCase)? */
static bool listing_in_suburb(const ca_re_board_t *b, const ca_re_listing_t *l,
                              const char *suburb) {
    for (size_t i = 0; i < b->p_count; ++i)
        if (cab_ord_eq(b->props[i].property_id, l->property_id))
            return cab_ci_eq(b->props[i].suburb, suburb);
    return false; /* property not found => no match (C# TryGetValue false) */
}

/* Stable descending sort of collected indices by ListedUtc. */
static void listing_sort_desc(const ca_re_board_t *b, size_t *idx, size_t n) {
    for (size_t i = 1; i < n; ++i) {
        size_t key = idx[i];
        int64_t kt = b->listings[key].listed_utc_ms;
        size_t j = i;
        while (j > 0 && b->listings[idx[j - 1]].listed_utc_ms < kt) {
            idx[j] = idx[j - 1];
            j--;
        }
        idx[j] = key;
    }
}

/* Collect the sorted active-in-suburb listing indices into *out_idx (owned).
 * Returns count, or SIZE_MAX on error/bad args. Sets *out_idx = NULL when 0. */
static size_t collect_active_in_suburb(const ca_re_board_t *b,
                                       const char *suburb, size_t **out_idx) {
    *out_idx = NULL;
    if (!b || cab_is_ws(suburb)) return (size_t)-1;
    if (b->l_count == 0) return 0;
    size_t *idx = (size_t *)malloc(b->l_count * sizeof(size_t));
    if (!idx) return (size_t)-1;
    size_t n = 0;
    for (size_t i = 0; i < b->l_count; ++i) {
        if (b->listings[i].is_active &&
            listing_in_suburb(b, &b->listings[i], suburb))
            idx[n++] = i;
    }
    if (n == 0) { free(idx); return 0; }
    listing_sort_desc(b, idx, n);
    *out_idx = idx;
    return n;
}

ca_re_listing_t *ca_re_board_active_in_suburb(const ca_re_board_t *b,
                                              const char *suburb,
                                              size_t *out_count) {
    if (!out_count) return NULL;
    size_t *idx = NULL;
    size_t n = collect_active_in_suburb(b, suburb, &idx);
    if (n == (size_t)-1) { *out_count = (size_t)-1; return NULL; }
    if (n == 0) { *out_count = 0; return NULL; }

    ca_re_listing_t *out = (ca_re_listing_t *)calloc(n, sizeof(*out));
    if (!out) { free(idx); *out_count = (size_t)-1; return NULL; }
    for (size_t i = 0; i < n; ++i) {
        if (!listing_copy(&out[i], &b->listings[idx[i]])) {
            ca_re_listing_free_array(out, i);
            free(idx);
            *out_count = (size_t)-1;
            return NULL;
        }
    }
    free(idx);
    *out_count = n;
    return out;
}

bool ca_re_board_suburb_average(const ca_re_board_t *b, const char *suburb,
                                ca_re_decimal_t *out) {
    if (out) *out = 0;
    if (!out) return false;
    size_t *idx = NULL;
    size_t n = collect_active_in_suburb(b, suburb, &idx);
    if (n == (size_t)-1 || n == 0) { free(idx); return false; }

    /* decimal Average => rounded mean of the micro-unit asking prices. */
    int64_t sum = 0;
    for (size_t i = 0; i < n; ++i) sum += b->listings[idx[i]].asking_price;
    free(idx);
    int64_t cnt = (int64_t)n;
    int64_t avg;
    if (sum >= 0) avg = (sum + cnt / 2) / cnt;
    else          avg = -((-sum + cnt / 2) / cnt);
    *out = (ca_re_decimal_t)avg;
    return true;
}

File: rust/src/real_estate.rs
//! real_estate — CircleAI real-estate-board primitives.
//!
//! Full Rust port of `src/CircleAI.RealEstate/RealEstatePrimitives.cs`:
//!
//! - Enum [`PropertyKind`] + records ([`Property`], [`Listing`], [`Valuation`],
//!   [`Viewing`]) + [`IRealEstateBoard`] with the deterministic in-memory
//!   [`InMemoryRealEstateBoard`] (property registry, listings + close,
//!   valuations, viewings, active-in-suburb + suburb-average comparable).
//!
//! `decimal` money maps to [`f64`].

use std::collections::HashMap;
use std::sync::Mutex;

use chrono::{DateTime, Utc};

/// (RealEstate) A property kind.
///
/// Mirrors `enum PropertyKind { Apartment, House, Townhouse, Commercial, Land }`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum PropertyKind {
    Apartment,
    House,
    Townhouse,
    Commercial,
    Land,
}

/// (RealEstate) A property.
///
/// Mirrors `sealed record Property(string PropertyId, string Suburb,
/// PropertyKind Kind, int Beds, int Baths, double FloorAreaM2)`.
#[derive(Debug, Clone, PartialEq)]
pub struct Property {
    pub property_id: String,
    pub suburb: String,
    pub kind: PropertyKind,
    pub beds: i32,
    pub baths: i32,
    pub floor_area_m2: f64,
}

impl Property {
    /// Constructs a property, mirroring the positional C# record constructor.
    pub fn new(
        property_id: impl Into<String>,
        suburb: impl Into<String>,
        kind: PropertyKind,
        beds: i32,
        baths: i32,
        floor_area_m2: f64,
    ) -> Self {
        Self {
            property_id: property_id.into(),
            suburb: suburb.into(),
            kind,
            beds,
            baths,
            floor_area_m2,
        }
    }
}

/// (RealEstate) A listing.
///
/// Mirrors `sealed record Listing(string ListingId, string PropertyId,
/// decimal AskingPrice, string Currency, DateTimeOffset ListedUtc,
/// bool IsActive)`.
#[derive(Debug, Clone, PartialEq)]
pub struct Listing {
    pub listing_id: String,
    pub property_id: String,
    pub asking_price: f64,
    pub currency: String,
    pub listed_utc: DateTime<Utc>,
    pub is_active: bool,
}

impl Listing {
    /// Constructs a listing, mirroring the positional C# record constructor.
    pub fn new(
        listing_id: impl Into<String>,
        property_id: impl Into<String>,
        asking_price: f64,
        currency: impl Into<String>,
        listed_utc: DateTime<Utc>,
        is_active: bool,
    ) -> Self {
        Self {
            listing_id: listing_id.into(),
            property_id: property_id.into(),
            asking_price,
            currency: currency.into(),
            listed_utc,
            is_active,
        }
    }
}

/// (RealEstate) A valuation.
///
/// Mirrors `sealed record Valuation(string PropertyId, decimal EstimatedValue,
/// string Source, DateTimeOffset AtUtc)`.
#[derive(Debug, Clone, PartialEq)]
pub struct Valuation {
    pub property_id: String,
    pub estimated_value: f64,
    pub source: String,
    pub at_utc: DateTime<Utc>,
}

impl Valuation {
    /// Constructs a valuation, mirroring the positional C# record constructor.
    pub fn new(
        property_id: impl Into<String>,
        estimated_value: f64,
        source: impl Into<String>,
        at_utc: DateTime<Utc>,
    ) -> Self {
        Self {
            property_id: property_id.into(),
            estimated_value,
            source: source.into(),
            at_utc,
        }
    }
}

/// (RealEstate) A scheduled viewing.
///
/// Mirrors `sealed record Viewing(string ViewingId, string ListingId,
/// string AttendeeName, DateTimeOffset AtUtc)`.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Viewing {
    pub viewing_id: String,
    pub listing_id: String,
    pub attendee_name: String,
    pub at_utc: DateTime<Utc>,
}

impl Viewing {
    /// Constructs a viewing, mirroring the positional C# record constructor.
    pub fn new(
        viewing_id: impl Into<String>,
        listing_id: impl Into<String>,
        attendee_name: impl Into<String>,
        at_utc: DateTime<Utc>,
    ) -> Self {
        Self {
            viewing_id: viewing_id.into(),
            listing_id: listing_id.into(),
            attendee_name: attendee_name.into(),
            at_utc,
        }
    }
}

/// (RealEstate) The real-estate board contract.
///
/// Mirrors `interface IRealEstateBoard`.
pub trait IRealEstateBoard {
    /// Registers (or overwrites) a property.
    fn register_property(&self, p: Property);
    /// Publishes (or overwrites) a listing.
    fn list(&self, l: Listing);
    /// Marks a listing inactive. Panics on an unknown listing id (mirrors the
    /// C# `InvalidOperationException`).
    fn close(&self, listing_id: &str);
    /// Records a valuation.
    fn value(&self, v: Valuation);
    /// Schedules a viewing.
    fn schedule_viewing(&self, v: Viewing);
    /// Active listings whose property is in `suburb` (case-insensitive), ordered
    /// by `listed_utc` descending.
    fn active_in_suburb(&self, suburb: &str) -> Vec<Listing>;
    /// The mean asking price of active listings in `suburb`, or `None` when there
    /// are none.
    fn suburb_average(&self, suburb: &str) -> Option<f64>;
}

/// (RealEstate) In-memory [`IRealEstateBoard`].
///
/// Mirrors `sealed class InMemoryRealEstateBoard`.
pub struct InMemoryRealEstateBoard {
    props: Mutex<HashMap<String, Property>>,
    listings: Mutex<HashMap<String, Listing>>,
    vals: Mutex<Vec<Valuation>>,
    viewings: Mutex<Vec<Viewing>>,
}

impl InMemoryRealEstateBoard {
    /// Creates an empty board.
    pub fn new() -> Self {
        Self {
            props: Mutex::new(HashMap::new()),
            listings: Mutex::new(HashMap::new()),
            vals: Mutex::new(Vec::new()),
            viewings: Mutex::new(Vec::new()),
        }
    }
}

impl Default for InMemoryRealEstateBoard {
    fn default() -> Self {
        Self::new()
    }
}

impl IRealEstateBoard for InMemoryRealEstateBoard {
    fn register_property(&self, p: Property) {
        self.props.lock().unwrap().insert(p.property_id.clone(), p);
    }

    fn list(&self, l: Listing) {
        self.listings.lock().unwrap().insert(l.listing_id.clone(), l);
    }

    fn close(&self, listing_id: &str) {
        let mut listings = self.listings.lock().unwrap();
        match listings.get(listing_id) {
            Some(l) => {
                let updated = Listing {
                    is_active: false,
                    ..l.clone()
                };
                listings.insert(listing_id.to_string(), updated);
            }
            None => panic!("Unknown listing {listing_id}"),
        }
    }

    fn value(&self, v: Valuation) {
        self.vals.lock().unwrap().push(v);
    }

    fn schedule_viewing(&self, v: Viewing) {
        self.viewings.lock().unwrap().push(v);
    }

    fn active_in_suburb(&self, suburb: &str) -> Vec<Listing> {
        if suburb.trim().is_empty() {
            panic!("suburb required");
        }
        let props = self.props.lock().unwrap();
        let listings = self.listings.lock().unwrap();
        let mut out: Vec<Listing> = listings
            .values()
            .filter(|l| {
                l.is_active
                    && props
                        .get(&l.property_id)
                        .is_some_and(|p| p.suburb.eq_ignore_ascii_case(suburb))
            })
            .cloned()
            .collect();
        // OrderByDescending(ListedUtc).
        out.sort_by(|a, b| b.listed_utc.cmp(&a.listed_utc));
        out
    }

    fn suburb_average(&self, suburb: &str) -> Option<f64> {
        let rows = self.active_in_suburb(suburb);
        if rows.is_empty() {
            return None;
        }
        let sum: f64 = rows.iter().map(|l| l.asking_price).sum();
        Some(sum / rows.len() as f64)
    }
}
